pytransform3d/rotations/_utils.py

```diff
@@ -59,8 +59,8 @@ def norm_angle(a):
     a_norm : float or array-like, shape (n,)
         Normalized angle(s) in radians
     """
-    # Source of the solution: http://stackoverflow.com/a/32266181
-    return -((np.pi - np.asarray(a)) % (2.0 * np.pi) - np.pi)
+    a = np.asarray(a, dtype=np.float64)
+    return (a - (np.ceil((a + np.pi) / (2.0 * np.pi)) - 1.0) * (2.0 * np.pi))
```

Comment:
@AlexanderFabisch Sorry for the late reply.
First I don't recommend using `norm_angle3(a)` as I mentioned before in the [discussion](https://github.com/dfki-ric/pytransform3d/discussions/260#discussioncomment-6514636). The behaviors between `%` and `numpy.fmod` are different and the result of `norm_angle3(a)` won't be as expected when `np.pi - np.asarray(a)` is negative: 
```Python
angle = 7.123
print(norm_angle1(angle))  # 0.839814692820414
print(norm_angle2(angle))  # 0.839814692820414
print(norm_angle3(angle))  # 7.123
print(norm_angle4(angle))  # 0.839814692820414
```
Second, by `numpy.finfo` we can get the `eps` and `epsneg`, which measures the smallest value can be added to 1.0 and be substracted from 1.0.
```cmd
>>> print(np.finfo(np.float64))  
Machine parameters for float64
---------------------------------------------------------------
precision =  15   resolution = 1.0000000000000001e-15
machep =    -52   eps =        2.2204460492503131e-16
negep =     -53   epsneg =     1.1102230246251565e-16
minexp =  -1022   tiny =       2.2250738585072014e-308
maxexp =   1024   max =        1.7976931348623157e+308
nexp =       11   min =        -max
smallest_normal = 2.2250738585072014e-308   smallest_subnormal = 4.9406564584124654e-324
---------------------------------------------------------------
```
It is important because the results returned by `%` operator and `numpy.remainder` used in `norm_angle1(a)` and `norm_angle4(a)` may not be represented exactly as a float when angle reaches the number near the eps(or epsneg) of `numpy.float64` and they will round it: 
```Python
# still valid
angle = -2.2204460492503131e-15
print(norm_angle1(angle))  # -2.220446049250313e-15
print(norm_angle2(angle))  # -2.220446049250313e-15
print(norm_angle3(angle))  # -2.220446049250313e-15
print(norm_angle4(angle))  # -2.220446049250313e-15

# invalid when it is near the eps or epsneg
angles = [1.1102230246251565e-16, 2.220446049250313e-16]
print(norm_angle1(angles))  # [-0. -0.]
print(norm_angle2(angles))  # [1.11022302e-16 2.22044605e-16]
print(norm_angle3(angles))  # [-0. -0.]
print(norm_angle4(angles))  # [-0. -0.]
```
However, when angle goes to extremely large, `norm_angle2(a)` will failed but `norm_angle1(a)` and `norm_angle4(a)` will be valid:
```Python
angles = 1e16
print(norm_angle1(angles))  # 1.7788350860040971
print(norm_angle2(angles))  # 2.0
print(norm_angle3(angles))  # 8.062020393183683
print(norm_angle4(angles))  # 1.7788350860040971
```
Based on my past experience, extremely large angle seldom(or never) occurred and precision is more important.

Lastly, if considering the convenience of rewriting in other programming languages(e.g. C/C++), I personally recommend not using `norm_angle1(a)` and `norm_angle4(a)` because `%` operator cannot handle float or double in C/C++ and the behaviors of `numpy.remainder` and C/C++'s `remainder` are different as described in the [doc of numpy.remainder](https://numpy.org/doc/stable/reference/generated/numpy.remainder.html). One may confused if he/she wants to rewrite the function in other programming languages. Below is the simple tests on [cppreference.com](https://en.cppreference.com/w/c/numeric/math/remainder):
**C**
```C
#include <stdio.h>
#include <math.h>
#include <fenv.h>

int main(void)
{
    printf("remainder(+5.1, +3.0) = %.1f\n", remainder(5.1,3));
    printf("remainder(-5.1, +3.0) = %.1f\n", remainder(-5.1,3));
    printf("remainder(+5.1, -3.0) = %.1f\n", remainder(5.1,-3));
    printf("remainder(-5.1, -3.0) = %.1f\n", remainder(-5.1,-3));
 
    // special values
    printf("remainder(+0.0, 1.0) = %.1f\n", remainder(+0.0, 1));
    printf("remainder(-0.0, 1.0) = %.1f\n", remainder(-0.0, 1));
    printf("remainder(+5.1, Inf) = %.1f\n", remainder(5.1, INFINITY));
 
    // error handling
    feclearexcept(FE_ALL_EXCEPT);
    printf("remainder(+5.1, 0) = %.1f\n", remainder(5.1, 0));
    if(fetestexcept(FE_INVALID)) puts("    FE_INVALID raised");
}
```
**Results in C**
```cmd
remainder(+5.1, +3.0) = -0.9
remainder(-5.1, +3.0) = 0.9
remainder(+5.1, -3.0) = -0.9
remainder(-5.1, -3.0) = 0.9
remainder(+0.0, 1.0) = 0.0
remainder(-0.0, 1.0) = -0.0
remainder(+5.1, Inf) = 5.1
remainder(+5.1, 0) = -nan
    FE_INVALID raised
```
**Python**
```Python
print("np.remainder(+5.1, +3.0) = %.1f\n" % np.remainder(5.1,3))
print("np.remainder(-5.1, +3.0) = %.1f\n" % np.remainder(-5.1,3))
print("np.remainder(+5.1, -3.0) = %.1f\n" % np.remainder(5.1,-3))
print("np.remainder(-5.1, -3.0) = %.1f\n" % np.remainder(-5.1,-3))

# special values
print("np.remainder(+0.0, 1.0) = %.1f\n" % np.remainder(0.0, 1))
print("np.remainder(-0.0, 1.0) = %.1f\n" % np.remainder(-0.0, 1))
print("np.remainder(+5.1, Inf) = %.1f\n" % np.remainder(5.1, np.inf))

# error handling
print("np.remainder(+5.1, 0) = %.1f\n" % np.remainder(5.1, 0))
```
**Results in Python**
```cmd
np.remainder(+5.1, +3.0) = 2.1
np.remainder(-5.1, +3.0) = 0.9
np.remainder(+5.1, -3.0) = -0.9
np.remainder(-5.1, -3.0) = -2.1

np.remainder(+0.0, 1.0) = 0.0
np.remainder(-0.0, 1.0) = 0.0
np.remainder(+5.1, Inf) = 5.1
RuntimeWarning: invalid value encountered in remainder
  print("np.remainder(+5.1, 0) = %.1f\n" % np.remainder(5.1, 0))
np.remainder(+5.1, 0) = nan
```

Btw, it is more convenient if one wants to normalize an angle in a 2π wide interval around a center value: 
```Python
TWO_PI = 2.0 * np.pi

def norm_angle2(a, center):
    a = np.asarray(a)
    return a - (np.ceil((a + np.pi - center) / TWO_PI) - 1.0) * TWO_PI

angle = 2.0 * np.pi
# normalize to (0, 2 * Pi]
print(norm_angle2(angle, np.pi))  # 6.283185307179586
# normalize to (-Pi, Pi]
print(norm_angle2(angle, 0.0))    # 0.0
```

Reply:
Thanks a lot for explaining the details!

Reply:
Feel free to raise any questions or point out any mistakes at any time!

Reply:
Hi @AlexanderFabisch,
Yesterday while writing the unit tests, I tried to do more analysis on numerical precision. I found out that there exists disadvantages of of using `numpy.ceil`. As I mentioned above, `%` operator and `numpy.remainder` will failed when angles are near around the eps or epsneg; and `numpy.ceil` will failed when angles go to be extremely large. So I just want to test how close to eps `%` operator and `numpy.remainder` can keep valid and how large `numpy.ceil` can maintain precision:
```Python
# how close to eps and epsneg
# slightly greater than 0
for i in range(np.finfo(np.float64).precision+1):
    angle = np.array([np.finfo(np.float64).eps * (10**i), np.finfo(np.float64).epsneg * (10**i)])
    print(f'[{str(i).zfill(2)}] raw angle: {angle}')
    print(f'   norm_angle1: {norm_angle1(angle) == angle}')
    print(f'   norm_angle2: {norm_angle2(angle) == angle}')
    print(f'   norm_angle4: {norm_angle4(angle) == angle}')
    print(f'---------------------------------------------')

# slightly smaller than 0
for i in range(np.finfo(np.float64).precision+1):
    angle = -np.array([np.finfo(np.float64).eps * (10**i), np.finfo(np.float64).epsneg * (10**i)])
    print(f'[{str(i).zfill(2)}] raw angle: {angle}')
    print(f'   norm_angle1: {norm_angle1(angle) == angle}')
    print(f'   norm_angle2: {norm_angle2(angle) == angle}')
    print(f'   norm_angle4: {norm_angle4(angle) == angle}')
    print(f'---------------------------------------------')
```
The results:
```cmd
[00] raw angle: [2.2204460492503131e-16 1.1102230246251565e-16]
   norm_angle1: [False False]
   norm_angle2: [ True  True]
   norm_angle4: [False False]
---------------------------------------------
[01] raw angle: [2.2204460492503131e-15 1.1102230246251565e-15]
   norm_angle1: [ True False]
   norm_angle2: [ True  True]
   norm_angle4: [ True False]
---------------------------------------------
[02] raw angle: [2.2204460492503131e-14 1.1102230246251565e-14]
   norm_angle1: [ True  True]
   norm_angle2: [ True  True]
   norm_angle4: [ True  True]
...
```
We can see that `%` operator and `numpy.remainder` only lose at most **2** digits of precision (**because `%` operator will round it**). However, `numpy.ceil` will lose more digits of precision as the angle going larger (**because as the angle going larger, it will lose more digits of `np.pi` when computing `a + np.pi`**):
```Python
# large positive number
for i in range(np.finfo(np.float64).precision+1):
    angle = 10**i
    print(f'[{str(i).zfill(2)}] raw angle: {angle}')
    print(f'   norm_angle1: {norm_angle1(angle)}')
    print(f'   norm_angle2: {norm_angle2(angle)}')
    print(f'   norm_angle4: {norm_angle4(angle)}')
    print(f'---------------------------------------------')

# large negative number
for i in range(np.finfo(np.float64).precision+1):
    angle = -(10**i)
    print(f'[{str(i).zfill(2)}] raw angle: {angle}')
    print(f'   norm_angle1: {norm_angle1(angle)}')
    print(f'   norm_angle2: {norm_angle2(angle)}')
    print(f'   norm_angle4: {norm_angle4(angle)}')
    print(f'---------------------------------------------')
```
The results:
```cmd
[00] raw angle: 1
   norm_angle1: 1.0
   norm_angle2: 1.0
   norm_angle4: 1.0
---------------------------------------------
[01] raw angle: 10
   norm_angle1: -2.5663706143591725
   norm_angle2: -2.5663706143591725
   norm_angle4: -2.5663706143591725
---------------------------------------------
[02] raw angle: 100
   norm_angle1: -0.5309649148733833
   norm_angle2: -0.5309649148733797
   norm_angle4: -0.5309649148733833
---------------------------------------------
...
[13] raw angle: 10000000000000
   norm_angle1: -0.29170481341523313
   norm_angle2: -0.29296875
   norm_angle4: -0.29170481341523313
---------------------------------------------
[14] raw angle: 100000000000000
   norm_angle1: -2.9257570164604694
   norm_angle2: -2.921875
   norm_angle4: -2.9257570164604694
---------------------------------------------
[15] raw angle: 1000000000000000
   norm_angle1: 2.1652724889850994
   norm_angle2: 2.125
   norm_angle4: 2.1652724889850994
---------------------------------------------
```
From the points of the easy portability to other languages, I still recommend using `numpy.ceil`. However, `numpy.ceil` can only keep the numerical precision only when the angle is at the scale of about `1e2`. I'm not sure whether to proceed with this commit change or not. Please let me know your thoughts if it's convenient. Thanks!

Reply:
The most common use case of this function is most likely for angles between -3 pi and 3 pi. If this behavior is properly documented, I don't see any issue with the ceil version. However, could you test this range explicitly?

We could do the following:
- generate a linspace-like array between (-pi, pi] (probably a good way to generate it is using -np.arange(-pi, pi, delta))
- in a for loop, add c = n * pi, where n in range(-3, 4)
- call norm_angle
- check the difference between the normalized angle and the angle before adding c

I assume the ceil version could be better around zero, but the remainder version could be better around 3 pi. We would need to quantify that.

Reply:
I tested the range from -10 * pi to 10 * pi, but I only tested the cases n * pi (when n is even). This is because when n is odd, the normalized angle will not be equal to the original angle.  
(For example, suppose the original angle is pi, and after adding 3pi to it, it becomes 4pi. When 4pi is normalized, it becomes 0 instead of pi.)
```Python
a_norm = np.linspace(np.pi, -np.pi, 1000000, endpoint=False, dtype=np.float64)[::-1]
for b in np.linspace(-10.0 * np.pi, 10.0 * np.pi, 11):
    a = a_norm + b
    norm1 = norm_angle1(a)
    norm2 = norm_angle2(a)
    norm4 = norm_angle4(a)

    assert_array_almost_equal(norm1, a_norm)
    assert_array_almost_equal(norm2, a_norm)
    assert_array_almost_equal(norm4, a_norm)

    assert_array_almost_equal(norm1, norm2)
    assert_array_almost_equal(norm2, norm4)

    print(f'[{b//np.pi}] length different from a_norm')
    print(f'(norm1, norm2, norm4)')
    print(f'({len(norm1[norm1 != a_norm])}, {len(norm2[norm2 != a_norm])}, {len(norm4[norm4 != a_norm])})')
    
    print(f'[{b//np.pi}] total difference from a_norm')
    print(f'(norm1, norm2, norm4)')
    print(f'({(norm1 - a_norm).sum()}, {(norm2 - a_norm).sum()}, {(norm4 - a_norm).sum()})')
    print('----------------------------------------------')
```
**Summary:**
1. The normalized angles returned by these three functions are almost equal to the original angles;
2. The normalized angles returned by these three functions are also almost equal to each other;
3. When n is -6.0, -4.0, or -2.0, the ceil version appears to have a smaller difference compared to the original angle. In other cases, there is no difference between these three versions.

Reply:
Sounds good. Then we should use the ceil version!

Reply:
**Todo(just for reminding):**
1. Use dtype=np.float64 for np.asarray
2. Add 2.0 * pi as TWO_PI in [_constants.py](http://constants.py/)
3. Add disscusion in function comment
4. Add extremely small numbers for tests
5. Rebase to current develop branch